### PR TITLE
Cross Validation

### LINKED ISSUE #327: Add a post processor to perform cross validation
--------
Issue Description
--------

New feature, add a new post processor to perform cross validation. This is useful for ROM train, and models selection.

##### Please attach the input file(s) that generate this error.  The simpler the input, the faster we can find the issue.

----------------
For Change Control Board: Issue Review
----------------
This review should occur before any development is performed as a response to this issue.
- [x] 1. Is it tagged with a type: defect or improvement? __Improvement, new feature__
- [x] 2. Is it tagged with a priority: critical, normal or minor? __Normal__
- [x] 3. If it will impact requirements or requirements tests, is it tagged with requirements? __N/A__
- [x] 4. If it is a defect, can it cause wrong results for users? If so an email needs to be sent to the users. __Not a defect__
- [x] 5. Is a rationale provided? (Such as explaining why the improvement is needed or why current code is wrong.)  __Yes__

-------
For Change Control Board: Issue Closure
-------
This review should occur when the issue is imminently going to be closed.
- [ ] 1. If the issue is a defect, is the defect fixed? ***improvement***
- [ ] 2. If the issue is a defect, is the defect tested for in the regression test system?  (If not explain why not.) ***improvement***
- [ ] 3. If the issue can impact users, has an email to the users group been written (the email should specify if the defect impacts stable or master)? ***improvement***
- [ ] 4. If the issue is a defect, does it impact the latest stable branch? If yes, is there any issue tagged with stable (create if needed)? ***n/a***
- [ ] 5. If the issue is being closed without a merge request, has an explanation of why it is being closed been provided? ***#331***

### PR DESCRIPTION
**Pull request by wangcj05**
--------
Pull Request Description
--------
##### What issue does this change request address? (Use "#" before the issue to link it, i.e., #42.)
Closes #327 

##### What are the significant changes in functionality due to this change request?
Add a new post processor to perform cross validation.

----------------
For Change Control Board: Change Request Review
----------------
The following review must be completed by an authorized member of the Change Control Board.
- [x] 1. Review all computer code.
- [x] 2. If any changes occur to the input syntax, there must be an accompanying change to the user manual and xsd schema. If the input syntax change deprecates existing input files, a conversion script needs to be added (see Conversion Scripts).
- [x] 3. Make sure the Python code and commenting standards are respected (camelBack, etc.) - See on the [wiki](https://github.com/idaholab/raven/wiki/RAVEN-Code-Standards#python) for details.
- [x] 4. Automated Tests should pass, including run_tests, pylint, manual building and xsd tests. If there are changes to Simulation.py or JobHandler.py the qsub tests must pass.
- [x] 5. If significant functionality is added, there must  be tests added to check this. Tests should cover all possible options.  Multiple short tests are preferred over one large test. If new development on the internal JobHandler parallel system is performed, a cluster test must be added setting, in <RunInfo> XML block, the node ```<internalParallel>``` to True.
- [x] 6. If the change modifies or adds a requirement or a requirement based test case, the Change Control Board's Chair or designee also needs to approve the change.  The requirements and the requirements test shall be in sync.
- [x] 7. The merge request must reference an issue.  If the issue is closed, the issue close checklist shall be done.
- [x] 8. If an analytic test is changed/added is the the analytic documentation updated/added?

### CODE REVIEW

**Comment by wangcj05** [comment]
@alfoa This is the initial implementation, and I believe a lot of things need to be changed. If you have time, please take a look and let me know if you want to change something.

**Comment by wangcj05** [comment]
@alfoa I will start to clean my branch, add more tests, and documentations.

**Comment by wangcj05** [comment]
@alfoa Do you have time to take a look at this PR?

**Comment by alfoa** [comment]
ok

File: framework/CrossValidations/Factory.py [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+#End compatibility block for Python 3-------------------------------------------
+
+################################################################################
+from utils import utils
+from .CrossValidation import CrossValidation
+from .SklCrossValidation import SciKitLearn
+## [ Add new class here ]
+################################################################################
+## Alternatively, to fully automate this file:
+# from OutStreamManagers import *
```
Comment (aalfonsi) [inline framework/CrossValidations/Factory.py]: we can delete this (or better link the description to this class)

File: tests/framework/PostProcessors/CrossValidations/gold/simpleCrossValidation/dumpMC.xml [hunk 2]
```diff
@@ -0,0 +1,370 @@
+<data name="dumpMC" type="Pointset"><input>x2,x1</input><output>ans2,ans</output><inputFilename>dumpMC.csv</inputFilename><metadata><SamplerType>array([u'MonteCarlo', u'MonteCarlo', u'MonteCarlo', u'MonteCarlo',
```
Comment (aalfonsi) [inline tests/framework/PostProcessors/CrossValidations/gold/simpleCrossValidation/dumpMC.xml]: do we really need this outputfile?

**Comment by aalfonsi** [comment]
Everything looks good except a couple of comments... 

Thanks

File: tests/framework/PostProcessors/CrossValidations/gold/simpleCrossValidation/dumpMC.xml (hunk 2 above)
Comment (wangcj05) [reply tests/framework/PostProcessors/CrossValidations/gold/simpleCrossValidation/dumpMC.xml]: We do not need this, I will delete it.

File: framework/CrossValidations/Factory.py (hunk 1 above)
Comment (wangcj05) [reply framework/CrossValidations/Factory.py]: Ok, I will delete this.

**Comment by wangcj05** [comment]
Hi @alfoa @joshua-cogliati-inl,
When I tried to add a map between the parameters, I found the 'cross_validation' module is still available in scikit learn version 0.18 and 0.19, but they are deprecated and will be removed in 0.20.

I think for now it is ok to just use the module 'cross_validation', and moved to 'model_selection' when our minimum requirement of scikit learn is 0.20. The other reason for this is that the call for this methods is changed in module 'model_selection'. Please let me know what do you think.

**Comment by joshua-cogliati-inl** [comment]
Looking at the differences, it looks like in the 0.18 model_selection the n parameter is automatically calculated from the data.  I think that could be easily calculated from the data if we are in 0.17.  Are there other changes that we can insulate the user from?  Basically, if cross_validation is 0.17 - 0.19, I am fine with using it, but if switching will cause RAVEN input file changes, I would like to insulate the user as much as possible.

**Comment by wangcj05** [comment]
Yes, the methods inside module 'cross_validation' are consistence in versions 0.17~0.19. We do not need to change the input file when switching. But from 0.18, the way to call the cross validation methods inside 'model_selection' module and 'cross_validation' module is different. That's why I suggest to use 'cross_validation' module.

**Comment by alfoa** [comment]
Agree with Conjiang

**Comment by joshua-cogliati-inl** [comment]
If we do not need to change the input when we switch to model_selection then I have no objection to using cross_validation now. (Agree with Congjian )

**Comment by wangcj05** [comment]
@joshua-cogliati-inl Sorry if I did not express myself well. We do not need to change the input across the versions when using 'cross_validation' module, but the problem is that this module will be removed in 0.20. There are some input differences between 'cross_validation' and 'model_selection' which will lead the change of raven input file. Besides, some method names are also changed between 'cross_validation' and 'model_selection' modules, for example method 'LabelKFold' changed to 'GroupKFold'. Moreover, if we use 'cross_validation', we can just call the method like 'KFold(n_folds=3)' to get the train and test indices. For the 'model_selection' module, we need to call method like 'kf=KFold(n_splits=3)', then 'kf.split(inputData)' to generate the indices. Please let me know what do you think.

**Comment by joshua-cogliati-inl** [comment]
@wangcj05 Tell me if I am wrong. So KFold would be able to work both with model_selection and cross_validation since the parameters are functionally the same, and the user input would not need to change.  However as I understand it you are saying that some of the other methods have different enough inputs that we would not easily be able to do a conversion like that.  Are any of the other methods needed now (or before we switch to sklearn 0.18?)?  If not, maybe we can just leave them undocumented, and after we switch to 0.18 we can document them then.

**Comment by joshua-cogliati-inl** [comment]
Also, I think this pull request can go in even if we will have to change the RAVEN inputs when we start supporting 0.20, but I would rather not have to change the input if not too much effort.

**Comment by wangcj05** [comment]
Ok, I made some changes to make sure the input will be consistence. Based on my understanding, when we switch to use 'model_selection', the input can be the same, we only need to make the parameter 'n' optional instead of required.

**Comment by wangcj05** [comment]
I think this PR is ready to go in. @alfoa @joshua-cogliati-inl Please let me know if you have any questions.

**Comment by aalfonsi** [comment]
----------------
For Change Control Board: Change Request Review
----------------
The following review must be completed by an authorized member of the Change Control Board.
- [x] 1. Review all computer code. ***ok***
- [x] 2. If any changes occur to the input syntax, there must be an accompanying change to the user manual and xsd schema. If the input syntax change deprecates existing input files, a conversion script needs to be added (see Conversion Scripts).  ***new feature. Documented in user manual***
- [x] 3. Make sure the Python code and commenting standards are respected (camelBack, etc.) - See on the [wiki](https://github.com/idaholab/raven/wiki/RAVEN-Code-Standards#python) for details.  ***ok***
- [x] 4. Automated Tests should pass, including run_tests, pylint, manual building and xsd tests. If there are changes to Simulation.py or JobHandler.py the qsub tests must pass.  ***ok***
- [x] 5. If significant functionality is added, there must  be tests added to check this. Tests should cover all possible options.  Multiple short tests are preferred over one large test. If new development on the internal JobHandler parallel system is performed, a cluster test must be added setting, in <RunInfo> XML block, the node ```<internalParallel>``` to True.  ***tests added***
- [x] 6. If the change modifies or adds a requirement or a requirement based test case, the Change Control Board's Chair or designee also needs to approve the change.  The requirements and the requirements test shall be in sync.  ***n/a***
- [x] 7. The merge request must reference an issue.  If the issue is closed, the issue close checklist shall be done.  ***#327***
- [x] 8. If an analytic test is changed/added is the the analytic documentation updated/added?  ***n/a***

**Comment by aalfonsi** [comment]
Checklist passed...

Merging...

**Pull request merged**